Src/Particle/AMReX_WriteBinaryParticleData.H

```diff
@@ -0,0 +1,688 @@
+                {
+                    HdrFile << which[j] << ' ' << count[j] << ' ' << where[j] << '\n';
+                }
+
+                if (gotsome && pc.doUnlink)
+                {
+                    // Unlink any zero-length data files.
+                    Vector<Long> cnt(nOutFiles,0);
+
+                    for (int i = 0, N=count.size(); i < N; i++) {
+                        cnt[which[i]] += count[i];
+                    }
+
+                    for (int i = 0, N=cnt.size(); i < N; i++)
+                    {
+                        if (cnt[i] == 0)
+                        {
+                            std::string FullFileName = NFilesIter::FileName(i, filePrefix);
+                            FileSystem::Remove(FullFileName);
+                        }
+                    }
+                }
+            }
+        }
+    }
+
+    if (ParallelDescriptor::IOProcessor())
+    {
+        HdrFile.flush();
+        HdrFile.close();
+        if ( ! HdrFile.good())
+        {
+            amrex::Abort("ParticleContainer::Checkpoint(): problem writing HdrFile");
+        }
+    }
+}
+
+template <class PC, EnableIf_t<IsParticleContainer<PC>::value, int> foo = 0>
+void WriteBinaryParticleDataAsync (PC const& pc,
+                                   const std::string& dir, const std::string& name,
+                                   const Vector<int>& write_real_comp,
+                                   const Vector<int>& write_int_comp,
+                                   const Vector<std::string>& real_comp_names,
+                                   const Vector<std::string>& int_comp_names)
+{
+    BL_PROFILE("WriteBinaryParticleDataAsync");
+    AMREX_ASSERT(pc.OK());
+
+    AMREX_ASSERT(sizeof(typename PC::ParticleType::RealType) == 4 ||
+                 sizeof(typename PC::ParticleType::RealType) == 8);
+
+    constexpr int NStructReal = PC::NStructReal;
+    constexpr int NStructInt  = PC::NStructInt;
+    constexpr int NArrayReal  = PC::NArrayReal;
+    constexpr int NArrayInt   = PC::NArrayInt;
+
+    const int MyProc = ParallelDescriptor::MyProc();
+    const int NProcs = ParallelDescriptor::NProcs();
+    const int IOProcNumber = NProcs - 1;
+
+    AMREX_ALWAYS_ASSERT(real_comp_names.size() == pc.NumRealComps() + NStructReal);
+    AMREX_ALWAYS_ASSERT( int_comp_names.size() == pc.NumIntComps() + NStructInt);
+
+    std::string pdir = dir;
+    if ( not pdir.empty() and pdir[pdir.size()-1] != '/') pdir += '/';
+    pdir += name;
+
+    if (MyProc == IOProcNumber)
+    {
+        if ( ! amrex::UtilCreateDirectory(pdir, 0755))
+        {
+            amrex::CreateDirectoryFailed(pdir);
+        }
+    }
+    ParallelDescriptor::Barrier();
+
+    Long total_np = 0;
+    Vector<Long> np_per_level(pc.finestLevel()+1);
+    Vector<Vector<Long> > np_per_grid(pc.finestLevel()+1);
+    for (int lev = 0; lev <= pc.finestLevel(); lev++)
+    {
+        np_per_grid[lev] = pc.NumberOfParticlesInGrid(lev);
```

Comment:
We could also do local reduction only and then do mpi_allreduce after the for loop.

Reply:
And for the following code in `NumberOfParticlesInGrid`
```    if (!only_local) {                                                                                                                                                                                     
        ParallelAllReduce::Sum(&nparticles[0], ngrids, ParallelContext::CommunicatorSub());                                                                                                         
    }
```
I think @mrowan137 has shown it's much faster to gather and broadcast.  @mrowan137 Can you comment on that?  There is a function for the gather `ParallelDescriptor::GatherLayoutDataToVector`.

Reply:
@atmyers if you look at `DistributionMapping::makeKnapSack (const LayoutData<Real>& rcost_local, ...` in AMReX_DistributionMapping.cpp you can see the alternate gather/broadcast sequence, it should probably be faster than the `ParallelAllReduce::Sum` here.